src/main/java/de/zalando/aruha/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -0,0 +1,366 @@
+            while (null != (toSend = e.getValue().takeEventsToStream(
+                    currentTimeMillis,
+                    Math.min(getParameters().batchLimitEvents, freeSlots),
+                    getParameters().batchTimeoutMillis))) {
+                flushData(e.getKey(), toSend);
+                this.sentEvents += toSend.size();
+                if (toSend.isEmpty()) {
+                    break;
+                }
+                freeSlots -= toSend.size();
+            }
+        }
+        pollPaused = getMessagesAllowedToSend() <= 0;
+        if (null != getParameters().batchKeepAliveIterations) {
+            final boolean allPartitionsReachedKeepAliveLimit = offsets.values().stream()
+                    .map(PartitionData::getKeepAlivesInARow)
+                    .allMatch(v -> v >= getParameters().batchKeepAliveIterations);
+            if (allPartitionsReachedKeepAliveLimit) {
+                shutdownGracefully();
+            }
+        }
+    }
+
+    private void flushData(final Partition.PartitionKey pk, final SortedMap<Long, String> data) {
+        final String evt = EventStream.createStreamEvent(
+                pk.partition,
+                String.valueOf(offsets.get(pk).getSentOffset()),
+                new ArrayList<>(data.values()),
+                Optional.empty());
+        try {
+            getOut().streamData(evt.getBytes(EventStream.UTF8));
+        } catch (final IOException e) {
+            LOG.error("Failed to write data to output. Session: " + getSessionId(), e);
+            shutdownGracefully();
+        }
+    }
+
+    @Override
+    public void onExit() {
+        if (null != topologyChangeSubscription) {
+            try {
+                topologyChangeSubscription.cancel();
+            } catch (final RuntimeException ex) {
+                LOG.warn("Failed to cancel topology subscription", ex);
+            } finally {
+                topologyChangeSubscription = null;
+                new HashSet<>(offsets.keySet()).forEach(this::removeFromStreaming);
+            }
+        }
+        if (null != kafkaConsumer) {
+            try {
+                kafkaConsumer.close();
+            } finally {
+                kafkaConsumer = null;
+            }
+        }
+    }
+
+    private void topologyChanged() {
+        if (!isCurrent()) {
+            return;
+        }
+        if (null != topologyChangeSubscription) {
+            topologyChangeSubscription.refresh();
+        }
+        reactNoTopologyChange();
+    }
+
+    private void reactNoTopologyChange() {
+        getZk().runLocked(() -> {
+            final Partition[] assignedPartitions = Stream.of(getZk().listPartitions())
+                    .filter(p -> getSessionId().equals(p.getSession()))
+                    .toArray(Partition[]::new);
+            addTask(() -> refreshTopologyUnlocked(assignedPartitions));
+        });
+    }
+
+    private void refreshTopologyUnlocked(final Partition[] assignedPartitions) {
```

Comment:
This name does not make it clear what this code is about. From my understanding this is about switching streaming partitions. Be it starting newly assigned partitions or releasing no longer assigned ones.

What about "reassignPartitionsAndUpdateConsumers"?